google_api_availability/lib/src/google_api_availability.dart

```diff
@@ -1,121 +1,100 @@
 import 'dart:async';
 import 'package:flutter/foundation.dart';
-import 'package:flutter/services.dart';
+import 'package:google_api_availability_platform_interface/google_api_availability_platform_interface.dart';
 
-import '../google_api_availability.dart';
-
-/// Flutter plugin to verify if Google Play Services are installed on the
-/// device.
+/// Flutter plugin to verify if Google Play Services are installed on the device.
 class GoogleApiAvailability {
   const GoogleApiAvailability._();
 
-  /// Creates an instance of the [GoogleApiAvailability] class. This
-  /// constructor is exposed for testing purposes only and should not be used
-  /// by clients of the plugin as it may break or change at any time.
+  /// Creates an instance of the [GoogleApiAvailability] class.
+  ///
+  /// This constructor is exposed for testing purposes only and should not be
+  /// used by clients of the plugin as it may break or change at any time.
   @visibleForTesting
   const GoogleApiAvailability.private();
 
   /// Acquires an instance of the [GoogleApiAvailability] class.
   static const GoogleApiAvailability instance = GoogleApiAvailability._();
```

Comment:
I am genuinly not sure about this, so just asking, but does GoogleApiAvailability._() equal null? Since you use that below to actually check both of them, why not just change it here to a nullable?

Reply:
I agree that this might be a little confusing; there is both `GoogleApiAvailability` and `GoogleApiAvailabilityPlatform`, both implementing the singleton pattern (having `instance`s). The `_()` and `private()` constructors of `GoogleApiAvailability` construct an instance of the class. In Dart, things starting with an `_` are considered private, and cannot be called by outsiders. This is great for implementing the singleton pattern, as we can now prevent the end user from accidentally instantiating the class multiple times. To make sure tests can still instantiate as much as they want, we expose the constructor `private()`, marked with `@visibleForTesting`.

The check we do at the method calls considers the instance of the platform interface, and is unrelated to `GoogleApiAvailability`.